src/math/range.cpp

```diff
@@ -3,45 +3,62 @@
 
 #include "math/range.h"
 
-Range::Range(double minimum, double maximum)
+Range::Range(std::optional<double> minimum, std::optional<double> maximum)
 {
-    minimum_ = minimum;
-    maximum_ = maximum;
+    minimum_ = std::min(minimum, maximum);
+    maximum_ = std::max(minimum, maximum);
 }
 
 /*
  * Data
  */
 
+// Set range
+void Range::set(std::optional<double> minimum, std::optional<double> maximum)
+{
+    minimum_ = minimum;
+    maximum_ = maximum;
+    if (isDefined() && minimum_ > maximum_)
+    {
+        std::swap(minimum_, maximum_);
+    }
+}
+
 // Set range
 void Range::set(double minimum, double maximum)
 {
     minimum_ = minimum;
     maximum_ = maximum;
+    if (minimum_ > maximum_)
+    {
+        std::swap(minimum_, maximum_);
+    }
 }
 
 // Set minimum for range
-void Range::setMinimum(double minimum) { minimum_ = minimum; }
+void Range::setMinimum(double minimum) { set(minimum, maximum_); }
 
 // Return minimum for range
-double Range::minimum() const { return minimum_; }
+double Range::minimum() const { return minimum_ ? *minimum_ : 0.0; }
 
 // Set maximum for range
-void Range::setMaximum(double maximum) { maximum_ = maximum; }
+void Range::setMaximum(double maximum) { set(minimum_, maximum); }
 
 // Return maximum for range
-double Range::maximum() const { return maximum_; }
+double Range::maximum() const { return maximum_ ? *maximum_ : 0.0; }
 
 // Return whether the range contains the specified value
 bool Range::contains(double d) const
 {
-    if ((d < minimum_) || (d > maximum_))
+    if (!isDefined() || !(d < minimum_) || (d > maximum_))
         return false;
     return true;
```

Comment:
You could one-line this if you wanted to:
```suggestion
    return isDefined() && (d >= minimum_) && (d <= maximum_);
```